File: php/src/Simple/InProgressOrder.php
<?php

namespace ImmutableDatamodel\Simple;

use DateTimeImmutable;
use ImmutableDatamodel\Shared\Address;
use ImmutableDatamodel\Shared\OrderConstraint;

readonly class InProgressOrder implements Order, Deliverable
{
    /**
     * @param string $orderId
     * @param OrderConstraint[] $constraints
     */
    public function __construct(
        private string $orderId,
        private array  $constraints
    ) {
    }

    public function orderId(): string
    {
        return $this->orderId;
    }

    public function constraints(): array
    {
        return $this->constraints;
    }

    public function deliver(Address $deliveryAddress, DateTimeImmutable $deliveryDate): DeliveringOrder
    {
        return new DeliveringOrder($this->orderId(), $this->constraints(), $deliveryAddress, $deliveryDate);
    }
}

File: golang/src/simple/in_progress_order.go
package simple

import (
	"example/shared"
	"time"
)

type InProgressOrder struct {
	orderId     string
	constraints []shared.OrderConstraint
}

func (ipo InProgressOrder) OrderId() string {
	return ipo.orderId
}

func (ipo InProgressOrder) Constraints() []shared.OrderConstraint {
	return ipo.constraints
}

func (ipo InProgressOrder) Deliver(deliveryAddress shared.Address, deliveryDate time.Time) (DeliveringOrder, error) {
	return NewDeliveringOrder(ipo, deliveryAddress, deliveryDate)
}
